numpy/lib/_function_base_impl.py

```diff
@@ -4355,12 +4355,14 @@ def quantile(a,
     The table above includes only the estimators from H&F that are continuous
     functions of probability `q` (estimators 4-9). NumPy also provides the
     three discontinuous estimators from H&F (estimators 1-3), where ``j`` is
-    defined as above and ``m`` and ``g`` are defined as follows.
+    defined as above, ``m`` is defined as follows, and ``g`` is a function
+    of the real-valued ``index = q*n + m - 1`` and ``j``.
 
-    1. ``inverted_cdf``: ``m = 0`` and ``g = int(q*n > 0)``
-    2. ``averaged_inverted_cdf``: ``m = 0`` and ``g = (1 + int(q*n > 0)) / 2``
+    1. ``inverted_cdf``: ``m = 0`` and ``g = int(index - j > 0)``
+    2. ``averaged_inverted_cdf``: ``m = 0`` and
+       ``g = (1 + int(index - j > 0)) / 2``
     3. ``closest_observation``: ``m = -1/2`` and
-       ``1 - int((g == 0) & (j%2 == 0))``
+       ``g = 1 - int((index == j) & (j%2 == 1))``
```

Comment:
Maybe I am not reading this correct, but the value `m=-1/2` does not make sense to me. Suppose we have `n=4` samples and `q=0`. Then `j = (q*n + m - 1) // 1 = (-3/2) // 1` so `j=-2`. The formula for the quantile `(1-g)*y[j] + g*y[j+1]` can then not be evaluated because both the indices `y[j]=y[-2]` and `y[j+1]=y[-1]` are not valid.
(note: also in the documentation for main I am struggling with this)

Reply:
Note that the docs state:
```
Note that indices j and j + 1 are clipped to the range 0 to n - 1 when the results
of the formula would be outside the allowed range of non-negative indices. 
The - 1 in the formulas for j and g accounts for Python’s 0-based indexing.
```

So with `q=0` you can generate a negative index and the function will return the bound.

The value `m=-1/2` is from Hyndmann and Fann. The method closest_neighbour applies round-to-even order behaviour when the real-valued index has a fractional value of 0.5. So with `m=-1/2` this case is identified as `j` is exactly integer. The rounding can then be applied appropriately if j is odd or even.

Reply:
What about `n=4`, `q=0.3`? Then we have `j=0.3*4-0.5-1 // 1 = = -1`. Both j and j+1 are clipped to zero. But the closest observation is the one corresponding to j=1

Reply:
Since q=0.3 does not create a fractional part of 0.5 on the real-valued index then the behaviour is unchanged from the current implementation:
```
>>> np.quantile([1, 2, 3, 4], 0.3, method='closest_observation')
1
```

Some quantile interpolation methods do not work very well on tiny arrays. The purpose of the PR is to clarify what the method is doing, not solve the inherent problems with it.

The reference implementation in R is the same:
```
> quantile(c(1, 2, 3, 4), 0.3, type=3)
30%
  1
```
With q=0.375 there is a difference when rounding the real valued index of 0.375*4=1.5:
```
>>> np.quantile([1, 2, 3, 4], 0.375, method='closest_observation')
1
>>> np.quantile([1, 2, 3, 4], 0.375001, method='closest_observation')
2
```

R:
```
> quantile(c(1, 2, 3, 4), 0.375, type=3)
37.5%
    2
```
Here R rounds up to the even order statistic, numpy rounds down to the odd order statistic (order is 1-based).